### PR TITLE
Autolabcli ng

### PR DESCRIPTION
**Pull request by 1AmNegan**
### Description of the Change

- Added skeleton structure to the project.

- Added "autolab init" command for POC.

- "autolab init" now takes now takes username and password as flags. If the flags are not provided, it will prompt the user to give the username and password. It prints back the username and password to the console.

- Added travis build and codecov integration.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env node
+
+let program = require('caporal');
```
Comment (prasadtalasila) [inline index.js]: It is better to push the caporal dependency and implementation details to lib/controller module. The index.js would only be responsible for configuring and using the objects exported out of lib/controller module.
For our case, the configuration could be just the logger configuration for caporal. The knowledge of utilizing caporal itself is hidden in lib/controller module. The index.js need not know the same.

File: lib/model/init.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+const inquirer = require('inquirer');
```
Comment (prasadtalasila) [inline lib/model/init.js]: The code provided in this file is still the CLI part. It is better to put this file as lib/cli/input/init.js. The code structure would be as follows.
```
lib/
    controller/
        init.js
    cli/
        input/
            init.js
        output/
             @init.js
```
For each command put in cli/input, there would be a matching prompt command. This structure should be completely testable. If you can propose a different structure that is more easier to test, I am willing to consider that too.

File: lib/cli/init.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const init = require('../model/init');
```
Comment (prasadtalasila) [inline lib/cli/init.js]: In the architecture, we have a clean separation between input/ and output/.

index.js calls caporal-based CLI controller structure put in place in lib/controller module. The controller module would call the relevant parts of the lib/cli/input module. The controller receives input objects, hands them over to model objects available in lib/model, receives outputs and passes them over to lib/cli/output.

File: lib/cli/init.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+module.exports = {
+   add: (program) => {
```
Comment (prasadtalasila) [inline lib/cli/init.js]: This caporal-specific configuration should be in lib/controller/init.js module.

File: lib/model/init.js (hunk 2 above)
Comment (prasadtalasila) [inline lib/model/init.js]: The lib/model module is to have data structures that help us complete the local computation if required. The lib/model does not deal with any input / output related work.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    "posttest": "nyc report --reporter=lcov > coverage.lcov && codecov"
+  },
+  "bin": {
+    "autolab": "./index.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/1AmNegan/autolabcli.git"
+  },
+  "author": "",
```
Comment (prasadtalasila) [inline package.json]: Please put in your name and modify the license to GPL v3.0

File: package.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  "scripts": {
+    "test": "nyc --reporter=lcov --reporter=text-lcov -s --all mocha 'test/**/*.js' > /dev/null",
+    "posttest": "nyc report --reporter=lcov > coverage.lcov && codecov"
+  },
+  "bin": {
+    "autolab": "./index.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/1AmNegan/autolabcli.git"
```
Comment (prasadtalasila) [inline package.json]: The URL needs to point to https://github.com/AutolabJS/autolabcli.git

Please correct at other places in the document too.

**Comment by prasadtalasila** [comment]
@AnkshitJain your inputs on the implementation skeleton suggested in this PR for the [autolabcli architecture](https://github.com/AutolabJS/autolabcli/wiki/AutolabcliNG) would be helpful as well.

File: lib/model/init.js (hunk 2 above)
Comment (1AmNegan) [reply lib/model/init.js]: I agree with the structure as it makes it easier to test.

**Comment by prasadtalasila** [comment]
@1AmNegan You git configuration might be faulty. Your username does not show up in the commits 9028e21 and bdd7fe3.
Please put in the tests for all the modules. Tests allow us to see the testability of the proposed architecture.

**Comment by prasadtalasila** [comment]
Please check the following options to test console output: [sinon console stubs](https://gyandeeps.com/console-stubbing/), [test-console](https://github.com/jamesshore/test-console), [std-mocks](https://github.com/neoziro/std-mocks)

File: test/test.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+var assert = require('assert');
```
Comment (prasadtalasila) [inline test/test.js]: @1AmNegan This test file does not make sense. You can remove the same.

File: test/lib/controller/test.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
+const chai = require("chai");
```
Comment (prasadtalasila) [inline test/lib/controller/test.js]: The convention you are using is to create a test module named `test.js` for a ES6 module named `index.js`. To be consistent with the naming scheme, it is better to call the test module as `index.js`.

File: test/test.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+const chai = require("chai");
```
Comment (prasadtalasila) [inline test/test.js]: please rename the file to `index.js`

File: lib/cli/input/init.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    else {
+      return 'Please enter your password';
+    }
+  }
+},
+];
+
+const getInput = (args, options) => {
+  if (typeof options.u == 'string' && typeof options.p == 'string')
+    return new Promise((resolve, reject) => resolve({username: options.u, password: options.p}));
```
Comment (prasadtalasila) [inline lib/cli/input/init.js]: To be consistent across the code base, it might be better to completely remove the promises and put in async / await functions.

File: lib/cli/input/init.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
+const inquirer = require('inquirer');
+
+const questions = [
```
Comment (prasadtalasila) [inline lib/cli/input/init.js]: It is better to use domain-relevant entity names. This array can appropriately be called **credentials**.

File: lib/cli/input/init.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    else {
+      return 'Please enter your password';
+    }
+  }
+},
+];
+
+const getInput = (args, options) => {
+  if (typeof options.u == 'string' && typeof options.p == 'string')
```
Comment (prasadtalasila) [inline lib/cli/input/init.js]: Instead of reducing the work to a low-level logical conditions, it might be better to have a function named `areCredentialsValid()` which would wrap the two conditions for now. The advantage of putting it in a function is to retain the flexibility to have policies on username and password strings later on.

File: lib/controller/index.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+const program = require('caporal');
+
+const init = require('./init');
```
Comment (prasadtalasila) [inline lib/controller/index.js]: Based on our reading of existing literature on ES6 ([exploring ES6](http://exploringjs.com/es6/)), we evolved the following template structure for a ES6 module.
```
imports
private functions (often async / await to make them asynchronous)
private classes
public classes
module.export of public classes
```
please see an example file named [status.js](https://github.com/AutolabJS/AutolabJS/blob/master/load_balancer/status.js) from AutolabJS project. 

I see the proposed structure to be more readable. If you don't have any objections to the given template, please adopt the same for this project as well.

One valid objection could be that a class is a extra paper work for code in input/, controller/ and output/ modules. We could get away with having a function there. Still, the other modules like model/ demand a class-based approach. 

In summary, the input/, controller/ and output/ modules can have the following structure.
```
imports
private functions (often async / await to make them asynchronous)
public functions
module.export of public functions defined earlier
```
Let's use the class-based template for code in model/ module.

File: test/lib/cli/output/init.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+const initOutput = require('../../../../lib/cli/output/init');
+
+const { expect } = chai;
+
+chai.use(sinonChai);
+chai.should();
+
+describe('For init output', () => {
+  it('log out the expected string', () => {
+    const logSpy = sinon.spy(console, 'log');
```
Comment (prasadtalasila) [inline test/lib/cli/output/init.js]: I am very happy that sinon could mock console. This solves a lot of problems.

File: lib/cli/output/init.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
```
Comment (prasadtalasila) [inline lib/cli/output/init.js]: Please explore the output formatting packages using in the autolabcli application and use the necessary ones. Once you select a few packages, please update the [notes on libraries](https://github.com/AutolabJS/autolabcli/wiki/Notes-on-Libraries) wiki page.

File: package.json [hunk 16]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "autolabcling",
```
Comment (prasadtalasila) [inline package.json]: The name should be `autolabjs`. This is to bring consistency in the names across all the projects.

File: package.json [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  "version": "1.0.0",
+  "description": "Codebase for backward incompatible revision of autolabcli v0.1.1.",
+  "main": "index.js",
+  "scripts": {
+    "test": "nyc --reporter=lcov --reporter=text-lcov -s --all mocha 'test/**/*.js' > /dev/null",
+    "posttest": "nyc report --reporter=lcov > coverage.lcov && codecov"
+  },
+  "bin": {
+    "autolab": "./index.js"
```
Comment (prasadtalasila) [inline package.json]: please change the name to `autolabjs`

File: test/lib/controller/test.js [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+chai.use(sinonChai);
+chai.should();
+
+describe('For controller entry point', () => {
+
+  it('should call the other controllers', () => {
+    const initAddToSpy = sinon.spy(initController, 'addTo');
+    program
+      .version('1.0.0')
+      .description('A Command Line Interface (CLI) for Autolab');
```
Comment (prasadtalasila) [inline test/lib/controller/test.js]: name to be changed to AutolabJS

File: test/test.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+const { expect } = chai;
+
+chai.use(sinonChai);
+chai.should();
+
+describe('For application entry point', () => {
+
+  it('should start the controller', () => {
+    const startSpy = sinon.spy(controller, 'start');
+    exec('npm link;autolab', (err, stdout, stdin) => {
```
Comment (prasadtalasila) [inline test/test.js]: command name to be changed to `autolabjs`

File: lib/cli/input/init.js [hunk 20]
```diff
@@ -0,0 +1,42 @@
+const inquirer = require('inquirer');
+
+const questions = [
+{
+  name: 'username',
+  type: 'input',
+  message: 'Enter your Autolab username:',
```
Comment (prasadtalasila) [inline lib/cli/input/init.js]: name to be changed to AutolabJS

File: lib/controller/index.js [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+
+module.exports = {
+  start: () => {
+    program
+      .version('1.0.0')
+      .description('A Command Line Interface (CLI) for Autolab');
```
Comment (prasadtalasila) [inline lib/controller/index.js]: name to be changed to AutolabJS

File: lib/controller/init.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+const onInit = async (args, options, logger) => {
+    const answer = await initInput.getInput(args, options);
+    const output = await initModel.getOutputString(answer.username, answer.password);
+    initOutput.send(output);
+}
+
+module.exports = {
+   addTo: (program) => {
+     program
+      .command('init', 'Initialize the directory as git repository and login to autolab system')
```
Comment (prasadtalasila) [inline lib/controller/init.js]: name to be changed to AutolabJS

File: lib/model/init.js [hunk 23]
```diff
@@ -1,9 +1,11 @@
-const getOutputString = (username, password) => {
-  return new Promise( (resolve, reject) =>
-    resolve(`Your username is: ${username}\nYour password is: ${password}`)
-  );
-}
+const chalk = require('chalk');
```
Comment (prasadtalasila) [inline lib/model/init.js]: model modules are not to be responsible for view / presentation / format tasks. It is better to export an object out of the lib/model/init.js module which will be used by the view to format and present.
The pattern of exporting data objects from model to views is pretty common. Please change the code accordingly.

File: lib/controller/index.js [hunk 24]
```diff
@@ -0,0 +1,15 @@
+const program = require('caporal');
+
+const init = require('./init');
+
+module.exports = {
+  start: () => {
```
Comment (prasadtalasila) [inline lib/controller/index.js]: please move the code out of module.exports into a public function which can be exported and called.

File: lib/controller/init.js [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+const onInit = async (args, options, logger) => {
+    initOutput.sendWelcome();
+    const answer = await initInput.getInput(args, options);
+    const output = await initModel.getOutputString(answer.username, answer.password);
+    initOutput.sendResult(output);
+}
+
+module.exports = {
+   addTo: (program) => {
```
Comment (prasadtalasila) [inline lib/controller/init.js]: please move the code out of module.exports into a public function

**Comment by prasadtalasila** [comment]
@1AmNegan The command name needs to be updated to 'autolabjs' in package.json

**Comment by prasadtalasila** [comment]
@1AmNegan Good work.
